[lldb/Target] Track containing StackFrameList to avoid circular dependencies

medismailben opened the pull request:
This change adds tracking of the StackFrameList that produced each frame by storing a weak pointer (m_frame_list_wp) in both `StackFrame` and `ExecutionContextRef`.

When resolving frames through `ExecutionContextRef::GetFrameSP`, the code now first attempts to use the remembered frame list instead of immediately calling `Thread::GetStackFrameList`. This breaks circular dependencies that can occur during frame provider initialization, where creating a frame provider might trigger `ExecutionContext` resolution, which would then call back into `Thread::GetStackFrameList()`, creating an infinite loop.

The `StackFrameList` now sets m_frame_list_wp on every frame it creates, and a new virtual method `GetOriginatingStackFrameList` allows frames to expose their originating list.

jimingham reviewed (approved):
This LGTM except for a suggested change in the comment for m_frame_list_wp.  You make it sound like this is some optional thing that you could choose to use or not.  But if it is resolveable it IS the stack frame list in which you are supposed to look up m_stack_id.  Your comment needs to say that.
Approved provided the comment is fixed.

medismailben committed:
[lldb/Target] Track containing StackFrameList to avoid circular dependencies

This change adds tracking of the StackFrameList that created each frame
by storing a weak pointer (m_frame_list_wp) in both StackFrame and
ExecutionContextRef.

When resolving frames through `ExecutionContextRef::GetFrameSP`, the code
now first attempts to use the remembered frame list instead of immediately
calling `Thread::GetStackFrameList`. This breaks circular dependencies that
can occur during frame provider initialization, where creating a frame
provider might trigger ExecutionContext resolution, which would then call
back into `Thread::GetStackFrameList`, creating an infinite loop.

The StackFrameList now sets m_frame_list_wp on every frame it creates,
and a new `GetContainingStackFrameList` method that allows frames to
expose their source list.

Signed-off-by: Med Ismail Bennani <ismail@bennani.ma>
